src/Features/CSharp/Portable/ExtractMethod/CSharpMethodExtractor.Analyzer.cs

```diff
@@ -10,79 +10,82 @@
-            return scope == null;
-        }
+            protected override bool ContainsReturnStatementInSelectedCode(IEnumerable<SyntaxNode> jumpOutOfRegionStatements)
+                => jumpOutOfRegionStatements.Where(n => n is ReturnStatementSyntax).Any();
 
-        protected override bool IsReadOutside(ISymbol symbol, HashSet<ISymbol> readOutsideMap)
-        {
-            if (!base.IsReadOutside(symbol, readOutsideMap))
-                return false;
+            protected override bool ReadOnlyFieldAllowed()
+            {
+                var scope = SelectionResult.GetContainingScopeOf<ConstructorDeclarationSyntax>();
+                return scope == null;
+            }
 
-            // Special case `using var v = ...` where the selection grabs the last statement that follows the local
-            // declaration.  The compiler here considers the local variable 'read outside' since it makes it to the
-            // implicit 'dispose' call that comes after the last statement.  However, as that implicit dispose would
-            // move if we move the `using var v` entirely into the new method, then it's still safe to move as there's
-            // no actual "explicit user read" that happens in the outer caller at all.
-            if (!this.SelectionResult.SelectionInExpression &&
-                symbol is ILocalSymbol { IsUsing: true, DeclaringSyntaxReferences: [var reference] } &&
-                reference.GetSyntax(this.CancellationToken) is VariableDeclaratorSyntax
-                {
-                    Parent: VariableDeclarationSyntax
-                    {
-                        Parent: LocalDeclarationStatementSyntax
-                        {
-                            Parent: BlockSyntax { Statements: [.., var lastBlockStatement] },
-                        },
-                    }
-                })
+            protected override bool IsReadOutside(ISymbol symbol, HashSet<ISymbol> readOutsideMap)
             {
-                var lastStatement = this.SelectionResult.GetLastStatement();
-                if (lastStatement == lastBlockStatement)
+                if (!base.IsReadOutside(symbol, readOutsideMap))
                     return false;
-            }
 
-            return true;
+                // Special case `using var v = ...` where the selection grabs the last statement that follows the local
+                // declaration.  The compiler here considers the local variable 'read outside' since it makes it to the
+                // implicit 'dispose' call that comes after the last statement.  However, as that implicit dispose would
+                // move if we move the `using var v` entirely into the new method, then it's still safe to move as there's
+                // no actual "explicit user read" that happens in the outer caller at all.
+                if (!this.SelectionResult.IsExtractMethodOnExpression &&
```

Comment:
renamed this property for consistency, and moved to an enum to fully specify it.